### ISSUE TITLE
SQLite database

### ISSUE THREAD

**Issue by ArtPoon** [PoonLab/covizu#81]
Currently we are storing unaligned and aligned sequences as flat files (FASTA).  The issue is that both header and sequence information is subject to change on the GISAID database, and we have to update our local data (#18).  Another issue is that our flat files are getting corrupted due to some sort of file writing error (#71).  

We can build a proper database with SQLite via the Python module fairly easily.  Suppose we implement the following schema:

* sequence
  * accession (primary key)
  * header
  * unaligned (original)
  * aligned (gotoh2)
* sample
  * accession
  * virus name
  * collection date
  * location
* sequencing
  * sequencing platform
  * assembly method
  * sample type (e.g., nasopharyngeal swab)
* acknowledgement
  * accession
  * originating lab
  * other...
* patient info
  * accession
  * gender
  * age
  * patient status

@ewong347 to review and edit

**Comment by ewong347** [comment]
Quick update:

db_utils.py contains wrapper functions for initializing & inserting into the sqlite database.
Autobot.py has been reworked to push into database, testing the script now– progress looking good but still a little slow (6 hours so far to align and insert 867/1621 seqs). 

Random idea:
With a database setup & the similarity of sequences, would it be worth it to hold a dictionary of some sort to hold the unaligned sequences with their aligned sequence counter part to speed up the alignment process? I think a string search (especially if we hashed the sequences) would be orders of magnitude quicker than the current procrustean alignment.

**Comment by ArtPoon** [comment]
Yes hashes would be an excellent idea!
We should consider running Autobot on multiple threads via MPI, which should be much easier on the cluster than your workstation.

**Comment by ewong347** [comment]
Running into issues with database locking when multi-threading. Need to find some solution to this. Stackoverflow not being too helpful. Will take another crack at it in the morning

```
sqlite3.OperationalError: database is locked
```

**Comment by ArtPoon** [comment]
Well, I'd rather prioritize having a proper database implemented soon over multithreading the alignment step.  What you propose seems reasonable to me (process in batch and then do a single transaction) but if it doesn't work out then let's backburner it.  Thanks!

**Comment by ArtPoon** [comment]
Check how many threads are being launched - alignment is memory intensive (>8GB per thread) and may lock up the computer.  Look up how `threading` is allocating resources.

**Comment by ArtPoon** [comment]
Make new table with parsed header information (see `filtering.py` for examples), such as sample collection dates.  This can also be used to regenerate `countries.json`.  Then `filtering.py` can query the database to apply our filters, *i.e.,* exclude records with missing dates.

**Comment by ewong347** [comment]
Looks like the spawned threads will automatically be balanced through the [OS](https://stackoverflow.com/questions/1787397/how-do-i-limit-the-number-of-active-threads-in-python). There is one thread spawned for each sequence in the handle, but these are tackled off of a queue.

I watched ```top``` for a bit - Memory usage ping pongs between 2-12 gb (peaking at 14% mem). CPU usage is quite high, but we're not even coming close to maxing out the memory.
I think the real time savings in the processing are coming from storing the aligned sequences with its unaligned hash. Benchmarked the ~400 sequences aligned in approximately 2h.

**Comment by ArtPoon** [comment]
Trim `N` (uncalled bases) from the 5' and 3' ends of unaligned sequences prior to calculating hash to reduce the number of redundant sequence records.

**Comment by ArtPoon** [comment]
`.strip('N')`

**Comment by ArtPoon** [comment]
When this stabilizes, please push to `dev` branch and we will work up a release to `master`

**Comment by ArtPoon** [comment]
I counted only about 42K sequences in the last fasta dump @ewong347 - is this right?
```console
art@Langley:/home/covid/covizu$ grep -c ">" data/gisaid-aligned.fa
42828
```
(modification date `Jun 27 00:43`)

**Comment by ewong347** [comment]
Looks correct to me. That's the same number of rows in the sqlite database. ChunkyBot is still running on the gsaid file in my local dev folder right now though, when it finishes it should catch all the sequences that were missed (should resolve the difference in count between the actual gisaid database and our own database).

**Comment by ArtPoon** [comment]
GISAID is showing 55K records, though.  Are we filtering at the export step?

**Comment by ewong347** [comment]
Found the issue: hashed sequences in the table didn’t reflect the change to the stripped version of the raw sequence. The chunky bot thought all the sequences were new, so was re-aligning all of them. Resettled the hashedseq table and re-running the script now. This iteration should be much quicker, and catch the ~8k missing sequences.

**Comment by ewong347** [comment]
Nested procceses were too slow (<1000 in 24 hours) - interrupted the process to align the missing sequences using the old update.py on a flat file, then will re-process the new alignment into the sqlite database.

**Comment by ewong347** [comment]
This problem was really bothering me so I did some investigation

Started by checking seqs in Database: 
```
ewong347@Langley:~/covizu$ grep -c '>' alignment.fa
>>42821
```

After running debug function built into Chunky bot (looking for missing sequences & updating flat file alignment with missing sequences), and subsequent update of aligned.fa

```
ewong347@Langley:~/covizu$ grep -c '>' aligned.fasta 
>>44132
```

We're still missing sequences here. So I re-downloaded all sequences using Chunkybot.py from gisaid website, this matches the counter on the website
```
ewong347@Langley:~/covizu$ grep -c '>' combined.fa
>>55477
```

Read in sequences in aligned.fa  & combined.fa. Using dictionary compares on accession number as dict key, print out sequences that are missing from alignment to missing.fa. Drop sequences less than 29000 nucleotides; this will be dropped by updater.py anyways 

552 sequences are missing from alignment due to len < 29000

``` 
ewong347@Langley:~/covizu$ grep -c '>' missing.fa
>> 12043
```

Raw Gisaid database - missing - pre-filtered == current database?
**BUT** 55477-12043-552 = 42,882  != 44132

There's is a discrepancy of ```1,250``` additional sequences in the sqlite database !!
There are *some* withdrawn sequences, to my knowledge this should only be in the low hundred's, not in the thousands.
Currently running update.py using sequences in the missing.fa file as the srcfile. 
I should get a better idea how many sequences were truly missing, vs. omitted by the aligning step after this process is complete. 

There are some sequences that I already know won't ever be aligned, even with the pre-filtering step, removing sequences <29,000 i.e. : 
```"Error in update_alignment: length of sequence x does not match reference."```

**Comment by ewong347** [comment]
dev@e2c6dae and dev@b2843aa addressed issues where ```ChunkyBot``` would iterate through the same fasta, resulting in uncaught missing sequences. These would be the 10k+ sequences that we are missing from the database. Current sqlite database has ```45,481/60,124``` live sequences, this has been exported to data/gisaid-aligned.fa. ```ChunkyBot``` is once again chugging away at the missing sequences. Due to issues with database locking, I will have ```ChunkyBot``` work on a database file in my home directory, then merge the databases once its run is complete.

An example of dev@e2c6dae in action:

```Sequence hCoV-19/Hong Kong/VB20079076/2020|EPI_ISL_482781|2020-06-01 with a length of 3822 cannot be aligned.```

**Comment by ArtPoon** [comment]
Based on the expected rate of progress (10s per genome) we should be up to date by now - did the script get killed at some point?

**Comment by ewong347** [comment]
Updater script still catching up on my local database file, there's about 3k sequences that still need to be caught up on (down from 8k, run time on chunkybot= 2,291:16). I've migrated all aligned the sequences in my local folder over to the database in the covid folder. There are now 52,949 sequences in the 'live' database out of the ~56k sequences that are of sufficient length on the gisaid platform.

**Comment by ArtPoon** [comment]
Ok thanks for the update -- there's really that many short sequences eh?  (nearly 10k)

**Comment by ArtPoon** [comment]
Reprocessing about 3,000 genomes consumed roughly 200 hours, which is an unusually long time given (conservatively) about 20 seconds per genome alignment.  Is this due to time-consuming database transactions?

**Comment by ArtPoon** [comment]
Do test run of autobot, timing every step including:
* download requested genomes of the day
* alignment to reference
* updating record in blank temporary database

**Comment by ewong347** [comment]
| Stage |Time stamp | diff from previous|
| ------------- | ------------- |-------------|
Script start | 11:21:30 |  0
Downloading complete | 11:23:09 | +1:39
Sequences inserted | 11:32:59 | + 9:50

I'd expect the download/ alignment to scale more or less linearly with time (given genome seqs are ~29k NT each). That being said, we might get lucky where a downloaded sequence is identical to one that has already been aligned. However there's 40k hashes, compared to 58k aligned sequences. We also might save on time if the sequence is shorter than ref.
 
Aligning 24 sequences at 9m50s = ~25 secs per on average

For a given insertion step (benchmarked all stages):

| Stage |Time stamp | diff from previous|
| ------------- | ------------- |-------------|
sequence handoff | 12:11:49 |  0
gotoh2 Aligner init | 12:11:49 | +0 (negligible)
Alignment complete | 12:12:13 | + 24 s
insert | 12:12:13| +0 (negligible)
commit | 12:12:13| +0 (negligible)

The 2.5 hour time that I mentioned in the dev meeting might just be when the mail utility sent me the crontab report- after the complete pipeline finished? I'll have to comfirm how I set it up. I'd be surprised if autobot (30 sequences added) took more than 15-20 minutes last night.
